src/trustRegion.jl

```diff
@@ -312,42 +324,55 @@ function perform_step!(cache::TrustRegionCache{false})
     cache.u_tmp = u .+ cache.step_size
     cache.fu_new = f(cache.u_tmp, p)
 
+    @unpack radius_update_scheme = cache
     trust_region_step!(cache)
     return nothing
 end
 
 function trust_region_step!(cache::TrustRegionCache)
-    @unpack fu_new, step_size, g, H, loss, max_trust_r = cache
+    @unpack fu_new, step_size, g, H, loss, max_trust_r, radius_update_scheme = cache
     cache.loss_new = get_loss(fu_new)
 
     # Compute the ratio of the actual reduction to the predicted reduction.
     cache.r = -(loss - cache.loss_new) / (step_size' * g + step_size' * H * step_size / 2)
     @unpack r = cache
 
-    # Update the trust region radius.
-    if r < cache.shrink_threshold
-        cache.trust_r *= cache.shrink_factor
-        cache.shrink_counter += 1
-    else
-        cache.shrink_counter = 0
-    end
-    if r > cache.step_threshold
-        take_step!(cache)
-        cache.loss = cache.loss_new
+    if radius_update_scheme === RadiusUpdateSchemes.Simple 
+      # Update the trust region radius.
+      if r < cache.shrink_threshold
+          cache.trust_r *= cache.shrink_factor
+          cache.shrink_counter += 1
+      else
+          cache.shrink_counter = 0
+      end
+      if r > cache.step_threshold
+          take_step!(cache)
+          cache.loss = cache.loss_new
 
-        # Update the trust region radius.
-        if r > cache.expand_threshold
-            cache.trust_r = min(cache.expand_factor * cache.trust_r, max_trust_r)
-        end
+          # Update the trust region radius.
+          if r > cache.expand_threshold
+              cache.trust_r = min(cache.expand_factor * cache.trust_r, max_trust_r)
+          end
+
+          cache.make_new_J = true
+      else
+          # No need to make a new J, no step was taken, so we try again with a smaller trust_r
+          cache.make_new_J = false
+      end
+
+      if iszero(cache.fu) || cache.internalnorm(cache.fu) < cache.abstol
+          cache.force_stop = true
+      end
+    
+    elseif radius_update_scheme === RadiusUpdateSchemes.Hei
```

Comment:
error for not implemented

Reply:
okay, like a TypeError when the `radius_update_scheme` type doesn't match `RadiusUpdateSchemes.T` ?